OsmAnd/src/net/osmand/plus/views/GPXLayer.java

```diff
@@ -213,36 +209,36 @@ public void onPrepareBufferImage(Canvas canvas, RotatedTileBox tileBox, DrawSett
 			}
 			drawSelectedFilesSplits(canvas, tileBox, selectedGPXFiles, settings);
 			drawSelectedFilesPoints(canvas, tileBox, selectedGPXFiles);
+			drawSelectedFilesStartEndPoints(canvas, tileBox, selectedGPXFiles);
 		}
 		if (textLayer != null && isTextVisible()) {
 			textLayer.putData(this, cache);
 		}
 
 	}
 
-	private int updatePaints(int color, boolean routePoints, boolean currentTrack, DrawSettings nightMode, RotatedTileBox tileBox) {
+	private int updatePaints(int color, String width, boolean routePoints, boolean currentTrack, DrawSettings drawSettings, RotatedTileBox tileBox) {
 		RenderingRulesStorage rrs = view.getApplication().getRendererRegistry().getCurrentSelectedRenderer();
-		final boolean isNight = nightMode != null && nightMode.isNightMode();
-		int hsh = calculateHash(rrs, routePoints, isNight, tileBox.getMapDensity(), tileBox.getZoom(),
-			currentTrackColorPref.get(), currentTrackWidthPref.get());
-		if (hsh != cachedHash) {
-			cachedHash = hsh;
+		boolean nightMode = drawSettings != null && drawSettings.isNightMode();
+		int hash = calculateHash(rrs, cachedTrackWidth, routePoints, nightMode, tileBox.getMapDensity(), tileBox.getZoom(),
+				currentTrackColorPref.get(), currentTrackWidthPref.get());
+		if (hash != cachedHash) {
+			cachedHash = hash;
 			cachedColor = ContextCompat.getColor(view.getApplication(), R.color.gpx_track);
+			defaultTrackWidth = 7 * view.getDensity();
```

Comment:
7?

Reply:
It was moved from here - https://github.com/osmandapp/Osmand/pull/9330/files#diff-3259c804faea84b62ab64ab1e0f87b01L277 for cases where track width parameter wasn't acquired from renderer or custom width has an invalid value.